keps/sig-node/xxxx-container-oom-kill-mode/README.md

```diff
@@ -0,0 +1,729 @@
+    OOMKillModeSingle OOMKillMode = "Single"
+
+    // OOMKillModeGroup kills all processes in the container when OOM is triggered
+    OOMKillModeGroup OOMKillMode = "Group"
+)
+```
+
+### Configuration Hierarchy
+
+The OOM kill behavior follows this precedence:
+
+1. **Container-level `oomKillMode`** (highest priority) - if explicitly set
+2. **Kubelet `singleProcessOOMKill` flag** - node-level default(Temporary measures during migration)
+3. **System default** - cgroup v2 defaults to group kill, cgroup v1 defaults to single process
+
+This hierarchy ensures backward compatibility while enabling smooth migration from node-level to container-specific configuration.
+
+### User Stories
+
+#### Story 1: Multi-process Web Application
+
+Multi-process web servers (PHP-FPM, Python Gunicorn/uWSGI, Ruby Unicorn) run a master process with multiple worker processes. When a worker consumes excessive memory due to a memory leak or processing a large request, only that problematic worker should be killed. This allows the application to continue serving requests with remaining healthy workers while the master spawns a replacement, ensuring high availability without full service disruption.
+
+#### Story 2: Database Container
+
+Databases like PostgreSQL require all processes to be killed together on OOM to ensure data consistency. Partial kills could leave the database in an inconsistent state.
+
+### Deprecation of singleProcessOOMKill Flag
+
+The `singleProcessOOMKill` kubelet flag will be removed in v1.38 (GA).
+
+#### Timeline
+
+- **v1.35 (Alpha)**: Deprecation announced in release notes and documentation
+- **v1.36 (Beta)**: Warning logs added when flag is used
+  ```
+  WARNING: Flag --single-process-oom-kill is deprecated and will be removed in v1.38.
+  Please migrate to container.oomKillMode field.
+  ```
+- **v1.38 (GA)**: Flag removed entirely
+
+#### Migration Guide
+
+Users currently using `--single-process-oom-kill` should:
+
+1. **Identify affected workloads**: Determine which containers rely on single-process OOM behavior
+2. **Add explicit configuration**: Set `oomKillMode: Single` for containers needing this behavior
+3. **Test migration**: Deploy with both flag and container field during transition
+4. **Remove flag**: Before v1.38, remove the kubelet flag from node configuration
+
+Example migration:
+
+```yaml
+# Before (v1.34)
+# kubelet --single-process-oom-kill=true
+
+# After (v1.35+)
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+    - name: multi-process-app
+      oomKillMode: Single # Explicitly set for containers needing it
+    - name: database
+      # No setting needed - will use default (Group)
+```
+
+### Notes/Constraints/Caveats
+
+- **cgroup v1 limitations**: On systems using cgroup v1, the `Group` mode cannot be enforced as cgroup v1 lacks the `memory.oom.group` mechanism. The API will accept the configuration but log a warning and fall back to single-process behavior.
```

Comment:
Should we fail clearly instead of logging a warning? If a cluster contains different node types (cgroup v1, cgroup v2, Windows), what's the impact if a pod gets scheduled onto a node that lacks the capacity required by the user config?

- A `oomKillMode = single / group` pod gets scheduled onto a Windows node.
- A `oomKillMode = group` pod gets scheduled onto a cgroup v1 node.